### PR TITLE
Edit Function Menu/Dialog (#528)

### PR DESCRIPTION
**Pull request by sajattack**
Not working yet. Putting this up for @xarkes to have a look and give me a hand.

### CODE REVIEW

File: src/dialogs/EditFunctionDialog.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#pragma once
```
Comment (xarkes) [inline src/dialogs/EditFunctionDialog.h]: Please use `#ifndef ... #define ... #endif` as in the other files.

File: src/dialogs/EditFunctionDialog.ui [hunk 2]
```diff
@@ -0,0 +1,96 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
+ <class>EditInstructionDialog</class>
```
Comment (xarkes) [inline src/dialogs/EditFunctionDialog.ui]: The class name is wrong here, it must be `EditFunctionDialog`

File: src/dialogs/EditFunctionDialog.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+EditFunctionDialog::EditFunctionDialog(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::EditFunctionDialog)
+{
+    ui->setupUi(this);
+    setWindowFlags(windowFlags() & (~Qt::WindowContextHelpButtonHint));
+
+    // Event filter for capturing Ctrl/Cmd+Return
+    ui->textEdit->installEventFilter(this);
```
Comment (xarkes) [inline src/dialogs/EditFunctionDialog.cpp]: There is no member called `textEdit` in the ui.

File: src/dialogs/EditFunctionDialog.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    close();
+}
+
+bool EditFunctionDialog::eventFilter(QObject */*obj*/, QEvent *event)
+{
+    if (event -> type() == QEvent::KeyPress) {
+        QKeyEvent *keyEvent = static_cast <QKeyEvent *> (event);
+
+        // Confirm comment by pressing Ctrl/Cmd+Return
+        if ((keyEvent -> modifiers() & Qt::ControlModifier) &&
```
Comment (xarkes) [inline src/dialogs/EditFunctionDialog.cpp]: Don't use spaces around `->`

**Comment by sajattack** [comment]
Thanks! Got it to open, now I just have to wire it up to the backend.

File: src/dialogs/EditFunctionDialog.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+void EditFunctionDialog::on_buttonBox_accepted()
+{
+}
+
+void EditFunctionDialog::on_buttonBox_rejected()
+{
+    close();
+}
+
+bool EditFunctionDialog::eventFilter(QObject */*obj*/, QEvent *event)
```
Comment (xarkes) [inline src/dialogs/EditFunctionDialog.cpp]: Better use Q_UNUSED than inline comment.

File: src/dialogs/EditFunctionDialog.h (hunk 1 above)
Comment (xarkes) [reply src/dialogs/EditFunctionDialog.h]: Push your new commits so I can actually see the changes you made.

File: src/menus/DisassemblyContextMenu.cpp [hunk 6]
```diff
@@ -681,6 +689,39 @@ void DisassemblyContextMenu::on_actionDeleteFunction_triggered()
...
+
+    if (dialog->exec()) {
+        QString new_name = dialog->getNameText();
+        Core()->renameFunction(fcn->name, new_name);
+        QString new_start_addr = dialog->getStartAddrText();
+        QString new_end_addr = dialog->getEndAddrText();
+        QString new_stack_size = dialog->getStackSizeText();
+        emit Core()->functionsChanged();
+    } else {
+        return;
```
Comment (xarkes) [inline src/menus/DisassemblyContextMenu.cpp]: Looks useless to me.

File: src/menus/DisassemblyContextMenu.cpp [hunk 7]
```diff
@@ -681,6 +689,39 @@ void DisassemblyContextMenu::on_actionDeleteFunction_triggered()
...
+    QString stackSizeText;
+    stackSizeText.sprintf("%d", fcn->stack);
+    dialog->setStackSizeText(stackSizeText);
+
+    if (dialog->exec()) {
+        QString new_name = dialog->getNameText();
+        Core()->renameFunction(fcn->name, new_name);
+        QString new_start_addr = dialog->getStartAddrText();
+        QString new_end_addr = dialog->getEndAddrText();
+        QString new_stack_size = dialog->getStackSizeText();
```
Comment (xarkes) [inline src/menus/DisassemblyContextMenu.cpp]: TODO Update the function with that, maybe check the commands under `af` (`af+`, `afb+`, ... or ask on irc)

File: src/Cutter.pro [hunk 8]
```diff
@@ -196,6 +196,7 @@ SOURCES += \
     widgets/RegisterRefsWidget.cpp \
     dialogs/SetToDataDialog.cpp \
     dialogs/SetFunctionVarTypes.cpp
```
Comment (xarkes) [inline src/Cutter.pro]: Missing backslash at end of line

File: src/Cutter.pro [hunk 9]
```diff
@@ -292,6 +293,7 @@ HEADERS  += \
     dialogs/SetToDataDialog.h \
     utils/InitialOptions.h \
     dialogs/SetFunctionVarTypes.h
```
Comment (xarkes) [inline src/Cutter.pro]: Missing backslash at end of line

File: src/menus/DisassemblyContextMenu.cpp [hunk 10]
```diff
@@ -681,6 +689,40 @@ void DisassemblyContextMenu::on_actionDeleteFunction_triggered()
...
+{
+    RCore *core = Core()->core();
+    EditFunctionDialog *dialog = new EditFunctionDialog(this);
+    RAnalFunction *fcn = r_anal_get_fcn_at (core->anal, offset, R_ANAL_FCN_TYPE_NULL);
+
+    dialog->setWindowTitle(tr("Edit function %1").arg(fcn->name));
+    dialog->setNameText(fcn->name);
+
+    QString startAddrText;
+    startAddrText.sprintf("0x%llx", fcn->addr);
```
Comment (radare) [inline src/menus/DisassemblyContextMenu.cpp]: %llx is not portable

File: src/menus/DisassemblyContextMenu.cpp [hunk 11]
```diff
@@ -681,6 +689,40 @@ void DisassemblyContextMenu::on_actionDeleteFunction_triggered()
...
+    dialog->setStartAddrText(startAddrText);
+
+    QString endAddrText;
+    endAddrText.sprintf("0x%llx", fcn->addr + fcn->_size);
```
Comment (radare) [inline src/menus/DisassemblyContextMenu.cpp]: llx not portable

File: src/menus/DisassemblyContextMenu.cpp (hunk 10 above)
Comment (sajattack) [reply src/menus/DisassemblyContextMenu.cpp]: What should I use instead? QString::number()?

Comment (xarkes) [reply src/menus/DisassemblyContextMenu.cpp]: Yep

**Comment by sajattack** [comment]
Can someone help me figure out why the Ok button isn't triggering the dialog->exec() if statement?

**Comment by ITAYC0HEN** [comment]
hi @sajattack ! :) I took a quick look and seems like you never defined a slots/signals for these buttons (OK | Cancel). This can be easily added from QT Creator:

image

Which will basically add these lines to the end of the UI file:

```
<connections>
  <connection>
   <sender>buttonBox</sender>
   <signal>accepted()</signal>
   <receiver>EditFunctionDialog</receiver>
   <slot>accept()</slot>
   <hints>
    <hint type="sourcelabel">
     <x>158</x>
     <y>183</y>
    </hint>
    <hint type="destinationlabel">
     <x>158</x>
     <y>101</y>
    </hint>
   </hints>
  </connection>
  <connection>
   <sender>buttonBox</sender>
   <signal>rejected()</signal>
   <receiver>EditFunctionDialog</receiver>
   <slot>reject()</slot>
   <hints>
    <hint type="sourcelabel">
     <x>158</x>
     <y>183</y>
    </hint>
    <hint type="destinationlabel">
     <x>158</x>
     <y>101</y>
    </hint>
   </hints>
  </connection>
 </connections>
</ui>
```

**Comment by jamieb122** [comment]
Just want to chime in on this one to see if function signature is in the works on this one as well?  Last I checked this may not be possible because it hasn't been implemented in r2, but this would be incredibly useful.

**Comment by sajattack** [comment]
@jamieb122 Let me get the basic features in the issue done first ;)

**Comment by radare** [comment]
What is not possible in r2? Define a function with from/to/stacksize?

**Comment by jamieb122** [comment]
> What is not possible in r2? Define a function with from/to/stacksize?

Define a function signature from prototype (i.e int cdecl func_name(int arg1, char* arg2)) that is then propagated to the names of local vars.  Last I looked I *believe* I saw the stub for it but it wasn't yet implemented.

**Comment by sajattack** [comment]
@jamieb122 That feels way out of scope of this PR to me.

**Comment by sajattack** [comment]
Only thing not working at the moment is the line that shows the end of the function isn't updating. Otherwise should be ready for review.

**Comment by ITAYC0HEN** [comment]
@jamieb122 -- it is indeed out of scope for this PR and we can take the discussion to our Telegram\IRC channel.
Eventually, defining a function by prototype would be implemented in a user-friendly way in Cutter.

**Comment by sajattack** [comment]
@ITAYC0HEN the only thing left is updating the lines that draw from the start to the end of a function. If you can let me know how to do that we're done I think.

**Comment by ITAYC0HEN** [comment]
can you please rebase your PR and resolve the conflicts? Then I'll fetch it locally on my side and try to help

**Comment by sajattack** [comment]
@ITAYC0HEN Done.

File: src/Cutter.pro [hunk 12]
```diff
@@ -200,6 +200,7 @@ SOURCES += \
     dialogs/SetFunctionVarTypes.cpp \
     widgets/ColorSchemePrefWidget.cpp \
     common/ColorSchemeFileSaver.cpp
```
Comment (xarkes) [inline src/Cutter.pro]: Still missing backslash which makes the build fail.

File: src/Cutter.pro [hunk 13]
```diff
@@ -300,6 +301,7 @@ HEADERS  += \
     dialogs/SetFunctionVarTypes.h \
     common/ColorSchemeFileSaver.h \
     widgets/ColorSchemePrefWidget.h
```
Comment (xarkes) [inline src/Cutter.pro]: Still missing backslash which makes the build fail.

**Comment by ITAYC0HEN** [comment]
@sajattack I checked the pr and seems like the feature works well. What "lines" do you refer to?

If you are talking about these lines, this is in radare2 and not in Cutter
image


If not these lines, so which

**Comment by sajattack** [comment]
@ITAYC0HEN Yeah the blue one in that picture.

**Pull request merged**